### ISSUE TITLE
how to take jax.jacfwd or rev on model when used Batch Norm

### ISSUE THREAD

**Issue by ak24watch** [patrick-kidger/equinox#1058]
@patrick-kidger 
@johannahaffner
@lockwo
@
**model should return vector to take jax.jaxfwd and rev** 

> when used batchnorm eqx suggest to update batch stats manually then return of model will be **(out_vector, stats)**

how to deal with it

**Comment by ak24watch** [comment]
can i use eqx.filter_jaxfwd

**Comment by A373k** [comment]
> how to use batchnorm stats when two or more call required to calculate loss  in single trian step

is it corect to use like this

```
dout, **bn_stats** = eqx.filter_grad(model)(x, bn_stats)
out, bn_stats = model(x, **bn_stats**)

```

**Comment by lockwo** [comment]
You can see the usage in the docs https://docs.kidger.site/equinox/examples/stateful/. It works the same for other autodiff operations.

```python
import equinox as eqx
import jax
import jax.numpy as jnp
import jax.random as jr
import optax  # https://github.com/deepmind/optax

# This model is just a weird mish-mash of stateful and non-stateful layers for
# demonstration purposes, it isn't doing any clever.
class Model(eqx.Module):
    norm1: eqx.nn.BatchNorm
    spectral_linear: eqx.nn.SpectralNorm[eqx.nn.Linear]
    norm2: eqx.nn.BatchNorm
    linear1: eqx.nn.Linear
    linear2: eqx.nn.Linear

    def __init__(self, key):
        key1, key2, key3, key4 = jr.split(key, 4)
        self.norm1 = eqx.nn.BatchNorm(input_size=3, axis_name="batch")
        self.spectral_linear = eqx.nn.SpectralNorm(
            layer=eqx.nn.Linear(in_features=3, out_features=32, key=key1),
            weight_name="weight",
            key=key2,
        )
        self.norm2 = eqx.nn.BatchNorm(input_size=32, axis_name="batch")
        self.linear1 = eqx.nn.Linear(in_features=32, out_features=32, key=key3)
        self.linear2 = eqx.nn.Linear(in_features=32, out_features=3, key=key4)

    def __call__(self, x, state):
        x, state = self.norm1(x, state)
        x, state = self.spectral_linear(x, state)
        x = jax.nn.relu(x)
        x, state = self.norm2(x, state)
        x = self.linear1(x)
        x = jax.nn.relu(x)
        x = self.linear2(x)
        return x, state

def compute_loss(model, state, xs, ys):
    batch_model = jax.vmap(
        model, axis_name="batch", in_axes=(0, None), out_axes=(0, None)
    )
    pred_ys, state = batch_model(xs, state)
    loss = jnp.mean((pred_ys - ys) ** 2)
    return loss, state


@eqx.filter_jit
def make_step(model, state, xs, ys):
    j, state = eqx.filter_jacfwd(compute_loss, has_aux=True)(model, state, xs, ys)
    return j, state



dataset_size = 10
seed = 5678

key = jr.PRNGKey(seed)
mkey, xkey, xkey2 = jr.split(key, 3)

model, state = eqx.nn.make_with_state(Model)(mkey)

xs = jr.normal(xkey, (dataset_size, 3))
ys = jnp.sin(xs) + 1

make_step(model, state, xs, ys)

```

**Comment by ak24watch** [comment]
> You can see the usage in the docs https://docs.kidger.site/equinox/examples/stateful/. It works the same for other autodiff operations.
> 
> import equinox as eqx
> import jax
> import jax.numpy as jnp
> import jax.random as jr
> import optax  # https://github.com/deepmind/optax
> 
> # This model is just a weird mish-mash of stateful and non-stateful layers for
> # demonstration purposes, it isn't doing any clever.
> class Model(eqx.Module):
>     norm1: eqx.nn.BatchNorm
>     spectral_linear: eqx.nn.SpectralNorm[eqx.nn.Linear]
>     norm2: eqx.nn.BatchNorm
>     linear1: eqx.nn.Linear
>     linear2: eqx.nn.Linear
> 
>     def __init__(self, key):
>         key1, key2, key3, key4 = jr.split(key, 4)
>         self.norm1 = eqx.nn.BatchNorm(input_size=3, axis_name="batch")
>         self.spectral_linear = eqx.nn.SpectralNorm(
>             layer=eqx.nn.Linear(in_features=3, out_features=32, key=key1),
>             weight_name="weight",
>             key=key2,
>         )
>         self.norm2 = eqx.nn.BatchNorm(input_size=32, axis_name="batch")
>         self.linear1 = eqx.nn.Linear(in_features=32, out_features=32, key=key3)
>         self.linear2 = eqx.nn.Linear(in_features=32, out_features=3, key=key4)
> 
>     def __call__(self, x, state):
>         x, state = self.norm1(x, state)
>         x, state = self.spectral_linear(x, state)
>         x = jax.nn.relu(x)
>         x, state = self.norm2(x, state)
>         x = self.linear1(x)
>         x = jax.nn.relu(x)
>         x = self.linear2(x)
>         return x, state
> 
> def compute_loss(model, state, xs, ys):
>     batch_model = jax.vmap(
>         model, axis_name="batch", in_axes=(0, None), out_axes=(0, None)
>     )
>     pred_ys, state = batch_model(xs, state)
>     loss = jnp.mean((pred_ys - ys) ** 2)
>     return loss, state
> 
> 
> @eqx.filter_jit
> def make_step(model, state, xs, ys):
>     j, state = eqx.filter_jacfwd(compute_loss, has_aux=True)(model, state, xs, ys)
>     return j, state
> 
> 
> 
> dataset_size = 10
> seed = 5678
> 
> key = jr.PRNGKey(seed)
> mkey, xkey, xkey2 = jr.split(key, 3)
> 
> model, state = eqx.nn.make_with_state(Model)(mkey)
> 
> xs = jr.normal(xkey, (dataset_size, 3))
> ys = jnp.sin(xs) + 1
> 
> make_step(model, state, xs, ys)

Thank you,
i got that i have seen it  ,what about this [question ](https://github.com/patrick-kidger/equinox/issues/1058#issuecomment-3139907124)

**Comment by lockwo** [comment]
that's fine [with aux=True]. read https://docs.kidger.site/equinox/examples/stateful/

**Comment by ak24watch** [comment]
> that's fine [with aux=True]. read https://docs.kidger.site/equinox/examples/stateful/
> 

I am asking is it correct to use batch stats twice for single train step, actually it's not question of equinox, but weird to know that  batch stats is been passed twice in single train step, which actually the implicit case like flax nnx
